pkg/provisioner/controller.go

```diff
@@ -425,14 +409,28 @@ func (c *obcController) handleProvisionClaim(key string, obc *v1alpha1.ObjectBuc
 	} else {
 		ob, err = c.provisioner.Grant(options)
 	}
+	// Record whether the provisioner returned an empty object bucket for error handling use later
+	emptyBucket := (ob == (&v1alpha1.ObjectBucket{}))
+
+	// Set information about the OB as soon as possible, even before handling errors so that the
+	// deferred cleanup function is not missing critical information needed to clean up.
+	// More note on above: can't check if provisioner returned empty OB after we set OB info below.
+	setObjectBucketName(ob, key)
+	ob.Spec.StorageClassName = obc.Spec.StorageClassName
+	if ob.Spec.ReclaimPolicy == nil || *ob.Spec.ReclaimPolicy == corev1.PersistentVolumeReclaimPolicy("") {
+		// Do not blindly overwrite the reclaim policy. The provisioner might have reason to
+		// specify a reclaim policy that is  different from the storage class.
+		ob.Spec.ReclaimPolicy = options.ReclaimPolicy
+	}
+	ob.SetLabels(c.provisionerLabels)
+
 	if err != nil {
 		return fmt.Errorf("error %s bucket: %v", verb, err)
-	} else if ob == (&v1alpha1.ObjectBucket{}) {
+	} else if emptyBucket {
 		return fmt.Errorf("provisioner returned empty object bucket")
 	}
 
-	// Fill in known information in the object bucket struct as soon as possible
-	setBasicOBInfo()
+	// This info set on the OB isn't critical for cleanup
 	ob.SetFinalizers([]string{finalizer})
 	ob.Spec.ClaimRef, err = claimRefForKey(key, c.libClientset)
 	if err != nil {
```

Comment:
I removed the `setBasicOBInfo` function in favor of doing this. This does not change the behavior but I think makes the code more conceptually linear. It does mean the `emptyBucket` check needs to be broken into two parts, and it does mean the error handling for `Provision()/Grant()` comes visually later in the code, but I believe those drawbacks are worth the mental linearity. I feel like this makes code maintenance easier for myself and others.